Enable GPU-based validation for Vulkan

ErichDonGubler opened the pull request:
**Connections**
_Link to the issues addressed by this PR, or dependent PRs in other repositories_

Resolves https://github.com/gfx-rs/wgpu/issues/1709.

- [x] Depends on https://github.com/gfx-rs/wgpu/pull/5146.
- [x] ~~Conflicts with https://github.com/gfx-rs/wgpu/pull/4297 without further API design (see [this conversation](https://github.com/gfx-rs/wgpu/pull/5046#discussion_r1468093046)).~~

**Description**
_Describe what problem this is solving, and how it's solved._

See <https://github.com/gfx-rs/wgpu/issues/1709>!

**Testing**
_Explain how this change is tested._

<!-- 
Thanks for filing! The codeowners file will automatically request reviews from the appropriate teams.

After you get a review and have addressed any comments, please explicitly re-request a review from the
person(s) who reviewed your changes. This will make sure it gets re-added to their review queue - you're no bothering us!
-->

**Checklist**

- [x] Run `cargo fmt`.
- [x] Run `cargo clippy`. If applicable, add:
  - ~~`--target wasm32-unknown-unknown`~~
  - ~~`--target wasm32-unknown-emscripten`~~
- [x] Run `cargo xtask test` to run tests.
- [x] Add change to `CHANGELOG.md`. See simple instructions inside file.

ErichDonGubler committed:
feat(vulkan): enable GPU-based validation for Vulkan backend

If [`VK_LAYER_KHRONOS_validation`] is present, and it supports
[`VK_EXT_validation_features`], we can configure it with another instance
creation info. element of type [`VkValidationFeaturesEXT`] to enable
GPU-based validation. Wire `InstanceFlags::GPU_BASED_VALIDATION` to do
this in the Vulkan backend. It's even already finding issues in our
`examples` and other tests! But…we'd like to handle those later, since
this is so important for users. So, I've broken that out to separate
issues. The instances we're aware of:

* `water` is running into sync. validation issues: see
  <https://github.com/gfx-rs/wgpu/issues/5231>
* `wgpu_test::shader::struct_layout::uniform_input` is failing to
    instrument shaders now; see
    <https://github.com/gfx-rs/wgpu/issues/5245>

It is apparent from this and the [DX12 implementation of GPU-based
validation] that we will need to communicate clearly to users that
`InstanceFlags::GPU_BASED_VALIDATION` implies
`InstanceFlags::VALIDATION`. Not all backends enforce this yet; I have
[split out this work][follow-up for flag implication].

Note that `VK_EXT_validation_features` has been deprecated in favor of
the more general layer configuration mechanism offered by
[`VK_EXT_layer_settings`].

[DX12 implementation of GPU-based validation]: https://github.com/gfx-rs/wgpu/pull/5146
[`VK_EXT_layer_settings`]: https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_EXT_layer_settings.html
[`VK_EXT_validation_features`]: https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_EXT_validation_features.html
[`VK_LAYER_KHRONOS_validation`]:https://vulkan.lunarg.com/doc/sdk/1.3.275.0/linux/khronos_validation_layer.html
[`VkValidationFeaturesEXT`]: https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkValidationFeaturesEXT.html
[follow-up for flag implication]: https://github.com/gfx-rs/wgpu/pull/5232

ErichDonGubler commented:
Linux CI is yielding errors like this ([link](https://github.com/gfx-rs/wgpu/actions/runs/7494262682/job/20402059107?pr=5046#step:13:8336)): `[2024-01-11T20:51:58Z ERROR wgpu_test::expectations] Unexpected failure due to: ValidationError(Some("Validation Error: [ VUID-VkDrawIndirectCommand-firstInstance-00501 ] Object 0: handle = 0x7fd64a7d1e00, type = VK_OBJECT_TYPE_QUEUE; | MessageID = 0x57f6b12e | vkQueueSubmit():  The drawIndirectFirstInstance feature is not enabled, but the firstInstance member of the VkDrawIndirectCommand structure at index 0 is not zero. The Vulkan spec states: If the drawIndirectFirstInstance feature is not enabled, firstInstance must be 0 (https://vulkan.lunarg.com/doc/view/1.3.268.0/linux/1.3-extensions/vkspec.html#VUID-VkDrawIndirectCommand-firstInstance-00501)"))
`

I haven't seen these in other PRs (CC @cwfitzgerald), so perhaps this extra validation is revealing an issue already? 🤔😀

cwfitzgerald commented:
@ErichDonGubler yeah that's absolutely an exposed bug!

cwfitzgerald commented on wgpu-hal/src/vulkan/instance.rs:
```diff
@@ -750,6 +750,17 @@ impl crate::Instance<super::Api> for super::Instance {
                 create_info = create_info.push_next(vk_create_info);
             }
 
+            let gpu_assisted_validation = &[vk::ValidationFeatureEnableEXT::GPU_ASSISTED];
+            let mut gpu_assisted_validation = vk::ValidationFeaturesEXT::builder()
+                .enabled_validation_features(gpu_assisted_validation)
+                .build();
+            if desc
+                .flags
+                .intersects(wgt::InstanceFlags::GPU_BASED_VALIDATION)
+            {
+                create_info = create_info.push_next(&mut gpu_assisted_validation);
+            }
```
I don't think we can add this without first verifying that the validation layers are there - vulkan requires that we only ask for what is available.

ErichDonGubler replied:
I'm not sure how to do this yet, since this is my first time working with Vulkan, but I'll give it a crack!

ErichDonGubler replied:
I _think_ I handled this with 17c854c3c.

teoxoy replied:
[`VK_EXT_validation_features`](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_EXT_validation_features.html) is the required extension to use this functionality.

It's an instance extension so we need to ask for it [here](https://github.com/gfx-rs/wgpu/blob/b7b7f7daed5defc85a8bc64b3cf16057c2175f59/wgpu-hal/src/vulkan/instance.rs#L227) and then check if the instance actually supports it. If it doesn't, we can't use the functionality.

It's worth noting that the extension is deprecated and was subsumed into [`VK_EXT_layer_settings`](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_EXT_layer_settings.html).

cwfitzgerald commented on wgpu-types/src/lib.rs:
```diff
@@ -890,6 +890,18 @@ bitflags::bitflags! {
+        ///   "GBV"](https://web.archive.org/web/20230206120404/https://learn.microsoft.com/en-us/windows/win32/direct3d12/using-d3d12-debug-layer-gpu-based-validation)
+        /// - Vulkan, via the `VK_LAYER_KHRONOS_validation` layer; called ["GPU-Assisted
+        ///   Validation"](https://github.com/KhronosGroup/Vulkan-ValidationLayers/blob/e45aeb85079e0835694cb8f03e6681fd18ae72c9/docs/gpu_validation.md#gpu-assisted-validation)
+        /// - Apple; called ["shader
```
We should note that we can't enable this, it's a environment variable thing.

ErichDonGubler replied:
Assuming this comment is about Apple's platform, I've removed this entirely. If you'd prefer to have it back with the note you've suggested, LMK, and I can add it back.

cwfitzgerald commented on wgpu-hal/src/vulkan/instance.rs:
```diff
@@ -750,6 +750,17 @@ impl crate::Instance<super::Api> for super::Instance {
+                .build();
+            if desc
+                .flags
+                .intersects(wgt::InstanceFlags::GPU_BASED_VALIDATION)
```
We should also consider enabling syncval

ErichDonGubler replied:
Uh...I actually don't know what that is! 👀💦

cwfitzgerald replied:
VK_VALIDATION_FEATURE_ENABLE_SYNCHRONIZATION_VALIDATION_EXT - we should probably use VK_VALIDATION_FEATURE_ENABLE_GPU_ASSISTED_EXT, VK_VALIDATION_FEATURE_ENABLE_GPU_ASSISTED_RESERVE_BINDING_SLOT_EXT, and VK_VALIDATION_FEATURE_ENABLE_SYNCHRONIZATION_VALIDATION_EXT

ErichDonGubler replied:
I _think_ I got this handled with bdfa467f7?

ErichDonGubler commented:
I've broken out the DX12 portion of this PR from here while I learn enough about Vulkan to implement @cwfitzgerald's feedback (see https://github.com/gfx-rs/wgpu/pull/5146). I've seen no controversy over the DX12 backend, and we on the Firefox team would like to take advantage of this as soon as we can.

ErichDonGubler commented:
Looks like we're running into yet more validation errors on Vulkan (CC @nical): https://github.com/gfx-rs/wgpu/actions/runs/7671904006/job/20911863256?pr=5046#step:13:1726

ErichDonGubler commented:
I noticed that the `PRINTF_SHADER` validation feature conflicts with `GPU_ASSISTED` (viz., the one this PR exposes):

https://github.com/KhronosGroup/Vulkan-ValidationLayers/blob/6ebabb71e297e0b8348ee4dd0633f2af3daaa771/docs/debug_printf.md#L174
 
I interpret this to mean that this change will conflict with #4297 (CC @exrook). We need to resolve this conflict with _some_ API design, but I have not yet determined a recommendation.

ErichDonGubler commented on wgpu-hal/src/vulkan/instance.rs:
```diff
@@ -756,6 +756,15 @@ impl crate::Instance<super::Api> for super::Instance {
                 create_info = create_info.push_next(vk_create_info);
             }
 
+            let mut gpu_assisted_validation = vk::ValidationFeaturesEXT::builder()
+                .enabled_validation_features(&[vk::ValidationFeatureEnableEXT::GPU_ASSISTED]);
+            if desc
+                .flags
+                .intersects(wgt::InstanceFlags::GPU_BASED_VALIDATION)
+            {
+                create_info = create_info.push_next(&mut gpu_assisted_validation);
+            }
+
```
I noticed that the `PRINTF_SHADER` validation feature conflicts with `GPU_ASSISTED` (viz., the one this PR exposes): https://github.com/KhronosGroup/Vulkan-ValidationLayers/blob/6ebabb71e297e0b8348ee4dd0633f2af3daaa771/docs/debug_printf.md#L174
 
I interpret this to mean that this change will conflict with #4297 (CC @exrook). We need to resolve this conflict with _some_ API design, but I have not yet determined a recommendation.

ErichDonGubler replied:
☝🏻CC @cwfitzgerald, @teoxoy.

cwfitzgerald replied:
Let's give priority to gpu based validation - if GBV is enabled, the printf feature isn't advertised.

ErichDonGubler replied:
@exrook: ☝🏻Then it's a race; whoever gets merged last to implement this! 😈🏁🏃💨

exrook commented:
RE: the CI failure, I'm hitting a failure on the same test in #4297, I haven't looked into it really but my suspicion is the validation layer output message is different from what the test harness expects to see.

cwfitzgerald commented:
Yeah - we expect a failure, which works. But then GPU validation cannot instrument the shader (as it's invalid spirv) and _that_ error is what fails the test.

ErichDonGubler committed:
docs: note that `GPU_BASED_VALIDATION` implies `VALIDATION`

ErichDonGubler commented on wgpu-types/src/lib.rs:
```diff
@@ -901,13 +901,15 @@ bitflags::bitflags! {
         const ALLOW_UNDERLYING_NONCOMPLIANT_ADAPTER = 1 << 3;
-        /// Enable GPU-based validation. Currently, this only changes behavior on the DX12
-        /// backend.
+        /// Enable GPU-based validation. Implies [`Self::VALIDATION`]. Currently, this only changes
```
@cwfitzgerald: FYI, added this `Implies …` statement to docs with a4e534979. I think this is correct (why have `GPU_BASED_VALIDATION` with `VALIDATION` off?), just wanted to call this out in case I'm misunderstanding something.

teoxoy commented:
The water example seems to fail now with:

```
[2024-02-02T20:25:58Z ERROR wgpu_test::expectations] Validation Error: Validation Error: [ SYNC-HAZARD-WRITE-AFTER-READ ] Object 0: handle = 0xa808d50000000033, type = VK_OBJECT_TYPE_RENDER_PASS; | MessageID = 0x376bc9df | vkCmdEndRenderPass: Hazard WRITE_AFTER_READ in subpass 0 for attachment 1 depth aspect during store with storeOp VK_ATTACHMENT_STORE_OP_STORE. Access info (usage: SYNC_LATE_FRAGMENT_TESTS_DEPTH_STENCIL_ATTACHMENT_WRITE, prior_usage: SYNC_FRAGMENT_SHADER_SHADER_SAMPLED_READ, read_barriers: VkPipelineStageFlags2(0), command: vkCmdDraw, seq_no: 2, reset_no: 1)
```

Which seems related to the `READ_ONLY_DEPTH_STENCIL` downlevel flag it's using.

https://github.com/gfx-rs/wgpu/issues/2243#issuecomment-984766597 is linking to some other Vulkan issues that are related.

cwfitzgerald commented:
I think it's acceptable to add that to the expected failures and fix the core problem in a followup.

ErichDonGubler committed:
refactor(vulkan): hoist validation layer check

This will be used shortly for checking if we should proceed with
enabling GPU-based validation.

ErichDonGubler commented on wgpu-hal/src/vulkan/instance.rs:
- [x] CI seems to be passing now, but that's unexpected. Definitely want to investigate before merging (see https://github.com/gfx-rs/wgpu/pull/5046#issuecomment-1927380045).

ErichDonGubler replied:
Looks like we probably need to be querying the layer's extensions, not just the instance's. Just validated that this works in CI! Will now fall back to needing to resolve CI. 🙂

ErichDonGubler replied:
Punted CI fixes to https://github.com/gfx-rs/wgpu/issues/5231 with 617c9e236cd3ac72ae093b011dbdae8bbb4f7e20. 👈🏻 CC @cwfitzgerald

ErichDonGubler committed:
refactor(vulkan): extract `enumerate_instance_extension_properties` helper

This will be used shortly for checking if we should proceed with
enabling GPU-based validation.

ErichDonGubler commented:
@teoxoy: Wanna "green" me, since I basically forced you to validate my code is correct while pairing? 😀

ErichDonGubler commented:
@cwfitzgerald, @teoxoy: I had to add another expected failure case to `wgpu_test::shader::struct_layout::uniform_input` to make CI green here; it accepts a failure to instrument shaders. I presume this is acceptable given that a follow-up issue is filed a la <https://github.com/gfx-rs/wgpu/issues/5231>?
